test/e2e/vmagent_test.go

```diff
@@ -291,3 +302,411 @@ func createExternalNodeCRD(data *TestData, nodeName string, ifName string, ip st
+	var err error
+	allURLs := append(append(allowedURLs, droppedURLs...), rejectedURLs...)
+	for _, url := range allURLs {
+		err := runCurlCommandOnVM(t, data, appliedToVM, url, crdv1alpha1.RuleActionAllow)
+		assert.Nil(t, err)
+	}
+
+	fqdnSettings := make(map[string]*crdv1alpha1.RuleAction, 0)
+	for _, url := range allowedURLs {
+		action := crdv1alpha1.RuleActionAllow
+		fqdnSettings[url] = &action
+	}
+	for _, url := range droppedURLs {
+		action := crdv1alpha1.RuleActionDrop
+		fqdnSettings[url] = &action
+	}
+	for _, url := range rejectedURLs {
+		action := crdv1alpha1.RuleActionReject
+		fqdnSettings[url] = &action
+	}
+
+	anp := createANPWithFQDN(t, data, name, namespace, appliedToVM, fqdnSettings)
+	for url, action := range fqdnSettings {
+		err = runCurlCommandOnVM(t, data, appliedToVM, url, *action)
+		assert.Nil(t, err)
+	}
+	err = data.DeleteANP(anp.Namespace, anp.Name)
+	require.Nil(t, err)
+	for _, url := range allURLs {
+		err := runCurlCommandOnVM(t, data, appliedToVM, url, crdv1alpha1.RuleActionAllow)
+		assert.Nil(t, err)
+	}
+	return nil
+}
+
+func testANPWithConnectivity(t *testing.T, data *TestData, name string, namespace string, appliedToVM vmInfo, peerVM vmInfo, proto AntreaPolicyProtocol, ingress bool, ruleAction crdv1alpha1.RuleAction, matchPeerEntity bool) error {
+	if proto != ProtocolTCP && proto != ProtocolUDP && proto != ProtocolICMP {
+		return fmt.Errorf("unsupported protocol %s when applying ANP rule to ExternalNode", proto)
+	}
+
+	var srcVM, dstVM vmInfo
+	if ingress {
+		srcVM = peerVM
+		dstVM = appliedToVM
+	} else {
+		srcVM = appliedToVM
+		dstVM = peerVM
+	}
+
+	var initialConnected, anpConnected bool
+	if ruleAction == crdv1alpha1.RuleActionDrop || ruleAction == crdv1alpha1.RuleActionReject {
+		initialConnected = true
+		anpConnected = false
+	} else if ruleAction == crdv1alpha1.RuleActionAllow {
+		initialConnected = false
+		anpConnected = true
+	}
+
+	if proto == ProtocolTCP || proto == ProtocolUDP {
+		err := runIperfServer(t, data, dstVM, iperfPort)
+		require.Nil(t, err, fmt.Sprintf("Failed to run iperf server on VM %s: %v", dstVM.nodeName, err))
+		defer func() {
+			assert.Nil(t, stopIperfCommand(t, data, dstVM))
+		}()
+	}
+
+	// The initial network connectivity is working as expected before ANP creation.
+	err := verifyANPConnectivity(t, data, srcVM, dstVM, initialConnected, proto, ruleAction)
+	require.Nil(t, err)
+	anp := createANP(t, data, name, namespace, ingress, proto, appliedToVM, peerVM, matchPeerEntity, ruleAction)
+	// The network connectivity is impacted by ANP.
+	err = verifyANPConnectivity(t, data, srcVM, dstVM, anpConnected, proto, ruleAction)
+	assert.Nil(t, err)
+
+	err = data.DeleteANP(anp.Namespace, anp.Name)
+	require.Nil(t, err, fmt.Sprintf("Failed to remove ANP %s: %v", name, err))
+	t.Logf("ANP test on name %s is done", name)
+	return nil
+}
+
+func createANP(t *testing.T, data *TestData, name, namespace string, ingress bool, proto AntreaPolicyProtocol,
```

Comment:
this is package scope, please name it more specifically, like createANPForExternalNode